Support pre-generated web thumbnails

po5 opened the pull request:
Supports YouTube videos and Twitch VODs. cf. #3, #111
No actual logic is specific to those sites, it's just that yt-dlp doesn't currently support anything else.
I may make PRs to yt-dlp later implementing "storyboards" for more websites.

It's good enough that I'd like others to try it and report issues.
Commits are a mess because I'm still experimenting a lot, they'll be squashed eventually.

I wonder if it may be possible to avoid spinning up a whole mpv process for each atlas, and instead use EDL or lavf:// to open multiple images and save them separately with an output template.

TODO:
- Reorganize the code, remove useless stuff
- Test with private videos (forward cookies setting)
- Better feedback if yt-dlp isn't found
- Mirror video filters to storyboard thumbnails (e.g. rotation, flipping, cropping)
- Even quicker thumbnail prioritization (by reserving n slots to the cursor?)
- Make it work nicely with the raw `thumbfast-render` API (even though there are currently no users of it)
- Make sure we never crash when switching between files
- Make sure we don't try to render thumbnails before they're written
- Make regular network thumbnailing placeholder consistent with this? (only broadcast info once we have the first thumbnail's dimensions, or the other way around, broadcast estimated dimensions before we get back the storyboard data)
- Make sure that absolutely everything is still fine with regular thumbnailing
- Fix audio crackling when spawning fragment downloads (happens even with --ao=null --ao-null-untimed) likely an mpv bug

My primary test videos:
- Short vid https://www.youtube.com/watch?v=HOpbPQGEH9k
- Long vid https://www.youtube.com/watch?v=fKM2NYABFFc
- Wrong thumbnail width reported by YouTube (159 instead of 160) https://www.youtube.com/watch?v=3To5DZ65n0U

po5 committed:
preliminary storyboard thumbnail implementation
last atlas index still has incorrectly sized thumbnails

po5 committed:
attempt at relative scaling
has wrong stride? wrong estimated number of thumbnails?

po5 committed:
fix scaling

po5 committed:
re-enable loop filter for storyboards

po5 committed:
fix y_start calc

po5 committed:
remove debug prints

po5 committed:
split stuff up better

po5 committed:
add prioritized process queue

po5 committed:
queued process cancellation, skip regular thumbnailer, fix stuck thumbs

po5 committed:
reset thumbnail_delta

po5 committed:
twitch vod support, refactored site matching, find ytdl path

po5 committed:
observe ytdl path

po5 committed:
cancellable ytdl storyboard json process and safety checks

po5 committed:
storyboard file deletion, regular thumbnailing fallback, hide stderr

po5 committed:
proper storyboard usage check

po5 committed:
recalculate storyboard thumbnail width

po5 committed:
exit out of watch_changes

Sneakpeakcss commented:
On Windows beside reverting https://github.com/po5/thumbfast/commit/9deb0733c4e36938cf90e42ddfb7a19a8b2f4641 i also had to remove `--vo=null` to get it working. 

Using `--vo=null` ends with an error: `thumbfast: storyboard download failed atlas: 2 status: 2`. The subprocess terminates without generating any `.bgra` files.

```
[   0.248][v][demux] Detected file format: webp_pipe (libavformat)
[   0.248][v][cplayer] Opening done: https://i.ytimg.com/sb/HOpbPQGEH9k/storyboard3_L3/M6.jpg?sqp=-oaymwENSDfyq4qpAwVwAcABBqLzl_8DBgjQ5MWpBg==&sigh=rs$AOn4CLAmRjZeNEAFVmD_SSdYuAV5WVIAiQ
[   0.248][v][cplayer] Running hook: auto_profiles/on_preloaded
[   0.248][v][cplayer] Failed sending hook command auto_profiles/on_preloaded. Removing hook.
[   0.248][v][lavf] select track 0
[   0.248][i][cplayer] ● Image  --vid=1  (webp 972x540)
[   0.248][f][cplayer] Error opening/initializing the selected video_out (--vo) device.
[   0.248][v][lavf] deselect track 0
[   0.248][i][cplayer] Video: no video
[   0.248][f][cplayer] No video or audio streams selected.
[   0.248][d][cplayer] Terminating demuxers...
[   0.248][d][ffmpeg] AVIOContext: Statistics: 41056 bytes read, 0 seeks
[   0.248][d][cplayer] Done terminating demuxers.
[   0.248][v][cplayer] finished playback, no audio or video data played (reason 4)
[   0.248][f][encode] no data written to target file
[   0.248][i][cplayer] Exiting... (Interrupted by error)
[   0.248][d][positioning] Destroying client handle...
[   0.248][d][cplayer] Run command: del, flags=64, args=[name="user-data/mpv/console"]
[   0.248][d][select] Destroying client handle...
[   0.248][d][commands] Destroying client handle...
[   0.249][d][stats] Destroying client handle...
[   0.249][d][console] Destroying client handle...
[   0.251][d][SystemMediaTransportControls] Destroying client handle...
```

Also about Twitch vods, i wonder if [this](https://github.com/mpv-player/mpv/issues/14840) won't be a potential issue here.

po5 committed:
remove vo=null

po5 commented:
Can reproduce, turns out in newer mpv versions (0.39+?) vo=null no longer works with the encoding mode? I removed it.
I'll look into the issue with 9deb0733c4e36938cf90e42ddfb7a19a8b2f4641 next time I boot into Windows.

> Also about Twitch vods, i wonder if https://github.com/mpv-player/mpv/issues/14840 won't be a potential issue here.

This is also a newly introduced issue, and it's completely irrelevant to us. They'll fix it eventually, probably.

Love running into new issues after updating mpv, wonder what else it broke (:
Edit: Looks like you can no longer set an empty value for `input-commands` in an auto-profile, mpv will straight up refuse to run. Fun.

po5 committed:
proper fix for miscalculated storyboard tile dimensions

po5 committed:
re-enable ytdl format selection for regular network thumbnailing

po5 committed:
use ytdl_hook json when available

po5 committed:
use extracted ytdl-compatible url

po5 committed:
fix ytdl process fallback and make it work in new mpv versions

po5 committed:
use closest available thumbnail index

Kladki commented on thumbfast.lua:
```diff
@@ -888,7 +1018,318 @@ local function sync_changes(prop, val)
+    referer_has_prefix = referer ~= "" and (referer_has_ytdl_prefix or referer_has_http_prefix or ytdl_opts.try_ytdl_first)
+
+    local checks = {
+        {input = path_ytdl,    patterns = youtube_patterns_free, condition = function(input) return true end},
+        {input = path,         patterns = youtube_patterns,      condition = function(input) return path_has_prefix end},
+        {input = path,         patterns = twitch_patterns,       condition = function(input) return path_has_prefix and string.match(input, twitch_base) end},
+        {input = referer_ytdl, patterns = youtube_patterns_free, condition = function(input) return input ~= "" end},
```
These `*_ytdl` variables just seem to be longer versions of the stripped ones, so why check against them?
```suggestion
        {input = path,    patterns = youtube_patterns_free, condition = function(input) return true end},
        {input = path,         patterns = youtube_patterns,      condition = function(input) return path_has_prefix end},
        {input = path,         patterns = twitch_patterns,       condition = function(input) return path_has_prefix and string.match(input, twitch_base) end},
        {input = referer, patterns = youtube_patterns_free, condition = function(input) return input ~= "" end},
```

po5 replied:
The freeform patterns in `youtube_patterns_free` match strings with anything at the start. Because the user may have prefixed the path with `ytdl://` to force ytdl on their end, which is an mpv invention that the yt-dlp command line doesn't understand, we have to make sure we're matching on the version on the path that we've stripped the ytdl:// prefix from.

For clarification, the `ytdl_prefix` pattern replace (`^ytdl://(.+)` -> `%1`) is used to get the path without this prefix while falling back to the full string if it didn't match.

Edit: Actually thinking about it some more I'm not sure I understand my own code anymore, why do I need to perform these matches on the version of the path that wasn't stripped from the http prefix? I'll look into it.

Kladki commented on thumbfast.lua:
```diff
@@ -888,7 +1018,318 @@ local function sync_changes(prop, val)
+                local sb_j = mp.utils.parse_json(sb_json.stdout)
+                if sb_j and sb_j.formats then
+                for _, sb in ipairs(sb_j.formats) do
+                if sb and sb.format_id == "sb0" and sb_j.duration and sb.width and sb.height and sb.fragments and #sb.fragments > 0 then
+                    local thumbnail_count = 0
+                    sb.rows = sb.rows or 5
+                    sb.columns = sb.columns or 5
```
Isn't the `sb.rows/shows or 5` redundant, since they are already checked to not be null here?

po5 replied:
Can you point out where you're saying they're checked to not be nil? I'm not seeing it.

Kladki replied:
Nevermind, I confused width and height for rows and columns. Should you even be defaulting to 5 instead of just not continuing if those attributes aren't set?

po5 replied:
Yeah fair.

Kladki commented on thumbfast.lua:
```diff
@@ -888,7 +1018,318 @@ local function sync_changes(prop, val)
+                        -- hack: youtube always adds 1 black frame at the end... --is this even true?
+                        if sb.extractor == "youtube" then
+                            thumbnail_count = thumbnail_count - 1
+                        end
```
It isn't, see https://www.youtube.com/watch?v=7Ing1lOzmGA for example (final storyboard is https://i.ytimg.com/sb/7Ing1lOzmGA/storyboard3_L3/M7.jpg?sqp=-oaymwENSDfyq4qpAwVwAcABBqLzl_8DBgjhmKSpBg==&sigh=rs$AOn4CLBgegFvp0Wrt7SIk5uTECSLgWEOWw)

Kladki commented on thumbfast.lua:
```diff
@@ -888,7 +1018,318 @@ local function sync_changes(prop, val)
+                    local storyboard_scale = {w=real_w/sb.width, h=real_h/sb.height}
+                    local thumbnail_size = {w=real_w, h=real_h}
+                    effective_w, effective_h = real_w, real_h
+                    info(real_w, real_h)
```
Scaling the storyboards to not fit the max should probably be an option, since it may not look good at some scaling ratios.

Kladki commented on thumbfast.lua:
```diff
@@ -888,7 +1018,318 @@ local function sync_changes(prop, val)
+    atlas:close()
+end
+
+local function fetch_fragment(storyboard, i, thumbnail_size, storyboard_scale)
+    if not storyboard.fragments[i] then return end
+
+    local args = {
+        mpv_path, storyboard.fragments[i].url, "--no-config", "--msg-level=all=no", "--really-quiet", "--no-terminal",
+        "--frames=1",
+        --"--load-scripts=no", "--osc=no", "--ytdl=no", "--load-stats-overlay=no", "--load-osd-console=no", "--load-auto-profiles=no",
+        "--no-sub", "--no-audio", "--hr-seek=no", "--sub-font-provider=none", "--embeddedfonts=no",
+        "--no-ytdl", "--demuxer-readahead-secs=0", "--demuxer-max-bytes=128KiB",
+        "--vd-lavc-software-fallback=1", "--vd-lavc-fast", "--vd-lavc-threads=2", --"--hwdec="..(options.hwdec and "auto" or "no"),
+        --"--vf="..vf_string(filters_all, true),
+        "--sws-allow-zimg=no", "--sws-fast=yes", "--sws-scaler=fast-bilinear",
+        --"--video-rotate="..last_rotate,
+        "--vf-add=format=bgra,scale=round(iw*"..storyboard_scale.w.."/"..thumbnail_size.w..")*"..thumbnail_size.w..":round(ih*"..storyboard_scale.h.."/"..thumbnail_size.h..")*"..thumbnail_size.h,
+        "--ovc=rawvideo", "--of=rawvideo", "--ofopts=update=1", "--o="..options.thumbnail..".ytdl"..tostring(i)
+    }
+
+    if os_name == "Mac" then
+        table.insert(args, "--macos-app-activation-policy=prohibited")
+    end
+
+    spawn_queued_process(args,
+        function(success, result, err)
+            if success == false or (not result or result.status ~= 0) then
+                if not result.killed_by_us then
+                    mp.msg.error("thumbfast: storyboard download failed", "atlas:", i, "status:", result.status)
+                end
+            else
+                get_thumb(options.thumbnail..".ytdl"..tostring(i), i, storyboard, thumbnail_size, storyboard_scale)
+            end
+        end
+    )
+
+    fetch_fragment(storyboard, i+1, thumbnail_size, storyboard_scale)
+end
```
Not too sure why you recurse here, when you could just have `fetch_fragments` just iterate over all indexes of `storyboard.fragments`

po5 replied:
It's a remnant from my old code thrown together to get stuff working, I'll clean it up for sure. :+1:

Kladki commented on thumbfast.lua:
```diff
@@ -888,7 +1018,318 @@ local function sync_changes(prop, val)
+    local stride = 4 * (thumbnail_size.w * math.min(storyboard.columns, atlas_pictures))
+    for pic = 0, atlas_pictures-1 do
+        local x_start = (pic % storyboard.columns) * thumbnail_size.w
+        local y_start = math.floor(pic / storyboard.columns) * thumbnail_size.h
```
Because you want to move down by one thumbnail when you reach the end of the row (works fine most of the time as-is though, since `columns` and `rows` are usually equal for most videos)
```suggestion
        local y_start = math.floor(pic / storyboard.rows) * thumbnail_size.h
```

po5 replied:
There will probably be more correctness errors like this in a future commit, I had to change the `get_thumb` function a bunch to support rotated, flipped, and cropped thumbnails. But that bug you pointed out still managed to survive through the rewrite, hahaha.

po5 replied:
The end of the row is reached when we exceed storyboard.columns. I'm pretty sure the current code is correct, and your suggestion is wrong. You'd even commented on 4b45873527b3dd8d6832dc27d0acd54476e5fcb9 letting me know I should use storyboard.columns.

Kladki replied:
Yeah, not sure what I was thinking...

Kladki commented on thumbfast.lua:
```diff
@@ -888,7 +1018,318 @@ local function sync_changes(prop, val)
+        local thumb_idx = (atlas_idx - 1) * storyboard.columns * storyboard.rows + pic
+        local filename = options.thumbnail .. ".ytdl-thumbx" .. tostring(thumb_idx)
+        local thumb_file = io.open(filename .. ".bgra", "wb")
+        for line = 0, thumbnail_size.h - 1 do
+            atlas:seek("set", 4 * x_start + (y_start + line) * stride)
+            local data = atlas:read(thumbnail_size.w * 4)
+            if data ~= nil then
+                thumb_file:write(data)
+            end
+        end
```
That seems like an awful lot of writes. Can't you just read the lines one by one, and keep appending them to a buffer, then write the complete buffer all at once?

N-R-K replied:
`io.open` very likely defaults to buffered io. Like most standard libraries in most languages.

Kladki commented on thumbfast.lua:
```diff
@@ -888,7 +1018,318 @@ local function sync_changes(prop, val)
+local ytdl_prefix = "^ytdl://(.+)"
+local subdomains = "[%w-.]*"
+local naked_ytdl_id = "^ytdl://([%w-_]+)$"
+local youtube_id = "[%w-_]+.*"
```
We know they are only ever 11 chars, and base64:
```suggestion
local youtube_id = "[%w-_]{11}.*"
```

po5 replied:
No such thing in Lua. Of course, we can repeat the character class 11 times instead. I'm just not too sure if there could be video IDs shorter than 11 characters that redirect to the canonical ID (or IDs where one of the 11 first characters is not one of the allowed ones but still resolves fine... I guess in that case the very first character could also be "wrong", we should just use .+ and call it a day), I know it happens if you add garbage at the end of a valid 11-char video ID. This costs nothing and is safer (: yt-dlp handles too-long video IDs just fine

Kladki commented on thumbfast.lua:
```diff
@@ -81,6 +81,95 @@ function subprocess(args, async, callback)
+    end
+end
+
+local function find_closest_index(tbl, target)
+    local lower = target - 1
+    local upper = target + 1
+
+    while lower >= 1 or upper <= #tbl do
+        if tbl[lower] ~= nil then
+            return lower
+        end
+        if tbl[upper] ~= nil then
+            return upper
+        end
+        lower = lower - 1
+        upper = upper + 1
+    end
+
+    return nil
+end
```
Shouldn't `tbl[target]` be checked at some point?
```suggestion
local function find_closest_index(tbl, target)
    if tbl[target] ~= nil then
        return target
    end

    local lower = target - 1
    local upper = target + 1

    while lower >= 1 or upper <= #tbl do
        if tbl[lower] ~= nil then
            return lower
        end
        if tbl[upper] ~= nil then
            return upper
        end
        lower = lower - 1
        upper = upper + 1
    end

    return nil
end
```

po5 replied:
It's checked before calling this function. This code is very ugly I'll think of another way to do it.

Kladki replied:
My bad then.

Kladki commented on thumbfast.lua:
```diff
@@ -888,7 +1018,318 @@ local function sync_changes(prop, val)
+    "^"..anycase("player%.twitch%.tv/.*[?&]video=v?")..twitch_id,
+}
+
+local function storyboard_supported_url(path, referer)
```
I also wonder if there even is any point having checks here as to whether the site supports yt-dlp, instead of just always attempting to fetch storyboards, and only continuing if they are returned.

po5 replied:
This would slow down the initial load of any non-storyboard network thumbnails, and cause extra requests because yt-dlp still has to fetch the list of all formats. It's not smart enough to lazily fetch format info based on `--format` and what's supported, which was a surprise to me. (and somewhat a blessing because it means that the existing json returned by ytdl_hook in newer mpv versions will always contain storyboard info, removing the need to spawn yt-dlp for those mpv versions)
I agree that this code is overengineered, but it's a port of the patterns from yt-dlp itself and should cover all cases.

po5 committed:
initial support for rotation and flipping
messy, and doesn't auto-update in all cases yet
crop handling is still missing
it's progress (:

po5 committed:
minor cleanup

po5 committed:
only fallback to preceeding thumbnail
to avoid spoilers

po5 committed:
staggered initial load for storyboard fragments

Commit:
fix local thumbnails after clearing preview
